GVFS/GVFS.Common/Database/GVFSDatabase.cs

```diff
@@ -0,0 +1,178 @@
+﻿using GVFS.Common.FileSystem;
+using GVFS.Common.Tracing;
+using System;
+using System.Collections.Concurrent;
+using System.Data;
+using System.IO;
+
+namespace GVFS.Common.Database
+{
+    /// <summary>
+    /// Handles setting up the database for storing data used by GVFS and
+    /// managing the connections to the database
+    /// </summary>
+    public class GVFSDatabase : IGVFSConnectionPool, IDisposable
+    {
+        private const int InitialPooledConnections = 5;
+        private const int MillisecondsWaitingToGetConnection = 50;
+
+        private bool disposed = false;
+        private string databasePath;
+        private IDbConnectionFactory connectionFactory;
+        private BlockingCollection<IDbConnection> connectionPool;
+
+        public GVFSDatabase(PhysicalFileSystem fileSystem, string enlistmentRoot, IDbConnectionFactory connectionFactory, int initialPooledConnections = InitialPooledConnections)
+        {
+            this.connectionPool = new BlockingCollection<IDbConnection>();
+            this.databasePath = Path.Combine(enlistmentRoot, GVFSConstants.DotGVFS.Root, GVFSConstants.DotGVFS.Databases.VFSForGit);
+            this.connectionFactory = connectionFactory;
+
+            string folderPath = Path.GetDirectoryName(this.databasePath);
+            fileSystem.CreateDirectory(folderPath);
+
+            for (int i = 0; i < initialPooledConnections; i++)
+            {
+                this.connectionPool.Add(this.connectionFactory.OpenNewConnection(this.databasePath));
+            }
+
+            this.Initialize();
+        }
+
+        public void Dispose()
+        {
+            if (this.disposed)
+            {
+                return;
+            }
+
+            this.disposed = true;
+            this.connectionPool.CompleteAdding();
+            while (!this.connectionPool.IsCompleted && this.connectionPool.TryTake(out IDbConnection connection))
+            {
+                connection.Dispose();
+            }

```

Comment:
Do we need to dispose the `connectionPool`? The documentation on [msdn](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.blockingcollection-1?view=netframework-4.8) seems to indicate that we should.